pkg/util/vsphere/vm/cdrom.go

```diff
@@ -0,0 +1,388 @@
+	var (
+		deviceChanges                  = make([]vimtypes.BaseVirtualDeviceConfigSpec, 0)
+		curCdromBackingFileNameToSpec  = make(map[string]vmopv1.VirtualMachineCdromSpec)
+		expectedBackingFileNameToCdrom = make(map[string]vimtypes.BaseVirtualDevice, len(vmCtx.VM.Spec.Cdrom))
+	)
+
+	for _, specCdrom := range vmCtx.VM.Spec.Cdrom {
+		imageRef := specCdrom.Image
+		if imageRef == nil {
+			return nil, fmt.Errorf("image ref is nil for CD-ROM: %q", specCdrom.Name)
+		}
+
+		cdrom, bFileName, err := getCdromAndBackingByImgRef(vmCtx, client, *imageRef, curDevices)
+		if err != nil {
+			vmCtx.Logger.Error(err, "failed to get CD-ROM device by image ref", "imageRef", *imageRef)
+			return nil, err
+		}
+
+		if cdrom != nil {
+			// CD-ROM already exists, collect it to update its connection state
+			// later with "Edit" operation.
+			curCdromBackingFileNameToSpec[bFileName] = specCdrom
+		} else {
+			// CD-ROM does not exist, create a new one with desired backing and
+			// connection, and update the device changes with "Add" operation.
+			cdrom = createNewCdrom(specCdrom, bFileName)
+			deviceChanges = append(deviceChanges, &vimtypes.VirtualDeviceConfigSpec{
+				Device:    cdrom,
+				Operation: vimtypes.VirtualDeviceConfigSpecOperationAdd,
+			})
+		}
+
+		expectedBackingFileNameToCdrom[bFileName] = cdrom
+	}
+
+	// Ensure all CD-ROM devices are assigned to controllers with proper keys.
+	// This may result new controllers to be added if none is available.
+	newControllers := ensureAllCdromsHaveControllers(expectedBackingFileNameToCdrom, curDevices)
+	deviceChanges = append(deviceChanges, newControllers...)
+
+	// Check and update existing CD-ROM devices' connection state if needed.
+	// Add them to the device changes with "Edit" operation.
+	curCdromChanges := updateCurCdromsConnectionState(
+		curCdromBackingFileNameToSpec,
+		expectedBackingFileNameToCdrom,
+	)
+	deviceChanges = append(deviceChanges, curCdromChanges...)
+
+	// Delete any existing CD-ROM devices that are not in the expected list.
+	// Add them to the device changes with "Remove" operation.
+	curCdroms := util.SelectDevicesByType[*vimtypes.VirtualCdrom](curDevices)
+	for _, cdrom := range curCdroms {
+		if b, ok := cdrom.Backing.(*vimtypes.VirtualCdromIsoBackingInfo); ok {
+			if _, ok := expectedBackingFileNameToCdrom[b.FileName]; ok {
+				continue
+			}
+		}
+
+		deviceChanges = append(deviceChanges, &vimtypes.VirtualDeviceConfigSpec{
+			Device:    cdrom,
+			Operation: vimtypes.VirtualDeviceConfigSpecOperationRemove,
+		})
+	}
+
+	return deviceChanges, nil
+}
+
+// UpdateConfigSpecCdromDeviceConnection updates the connection state of the
+// VM's existing CD-ROM devices to match what specifies in VM.Spec.Cdrom list.
+func UpdateConfigSpecCdromDeviceConnection(
+	vmCtx pkgctx.VirtualMachineContext,
+	client ctrlclient.Client,
+	config *vimtypes.VirtualMachineConfigInfo,
+	configSpec *vimtypes.VirtualMachineConfigSpec) error {
+	var (
```

Comment:
Please leave a blank line between the function signature and the `var` statement.